charon/src/ast/expressions_utils.rs

```diff
@@ -19,3 +20,93 @@ impl BorrowKind {
         }
     }
 }
+
+impl Place {
+    /// Compute the type of a place.
+    pub fn compute_projected_type(
+        &self,
+        type_decls: &Vector<TypeDeclId, TypeDecl>,
+        locals: &Vector<VarId, Var>,
+    ) -> Result<Ty, ()> {
+        // Lookup the local
+        let mut cur_ty = locals.get(self.var_id).ok_or(())?.ty.clone();
+
+        // Apply the projection
+        use ProjectionElem::*;
+        for pe in &self.projection {
+            cur_ty = match pe {
+                Deref => {
+                    use TyKind::*;
+                    match cur_ty.kind() {
+                        Ref(_, ty, _) | RawPtr(ty, _) => ty.clone(),
+                        Adt(..) | TypeVar(_) | Literal(_) | Never | TraitType(..) | DynTrait(_)
+                        | Arrow(..) => {
+                            // Unreachable
+                            return Err(());
+                        }
```

Comment:
To be fully correct we should handle boxes here too

Reply:
I updated the code